main/lsp/LSPPreprocessor.h

```diff
@@ -69,27 +51,27 @@ class LSPPreprocessor final {
     std::unique_ptr<LSPMessage> makeAndCommitWorkspaceEdit(std::unique_ptr<SorbetWorkspaceEditParams> params,
                                                            std::unique_ptr<LSPMessage> oldMsg);
 
-    /* The following methods convert edits into LSPFileUpdates. */
+    /* The following methods convert edits into SorbetWorkspaceEditParams. */
 
-    void canonicalizeEdits(u4 v, std::unique_ptr<DidChangeTextDocumentParams> changeParams,
-                           LSPFileUpdates &updates) const;
-    void canonicalizeEdits(u4 v, std::unique_ptr<DidOpenTextDocumentParams> openParams, LSPFileUpdates &updates) const;
-    void canonicalizeEdits(u4 v, std::unique_ptr<DidCloseTextDocumentParams> closeParams,
-                           LSPFileUpdates &updates) const;
-    void canonicalizeEdits(u4 v, std::unique_ptr<WatchmanQueryResponse> queryResponse, LSPFileUpdates &updates) const;
-    void mergeEdits(LSPFileUpdates &to, LSPFileUpdates &from);
+    std::unique_ptr<SorbetWorkspaceEditParams>
+    canonicalizeEdits(u4 v, std::unique_ptr<DidChangeTextDocumentParams> changeParams);
+    std::unique_ptr<SorbetWorkspaceEditParams> canonicalizeEdits(u4 v,
+                                                                 std::unique_ptr<DidOpenTextDocumentParams> openParams);
+    std::unique_ptr<SorbetWorkspaceEditParams>
+    canonicalizeEdits(u4 v, std::unique_ptr<DidCloseTextDocumentParams> closeParams);
+    std::unique_ptr<SorbetWorkspaceEditParams>
+    canonicalizeEdits(u4 v, std::unique_ptr<WatchmanQueryResponse> queryResponse) const;
 
     /**
-     * Returns a global state for typechecking, cloned from initialGS. Note: The clone does not share an error queue
-     * with initialGS.
+     * Get the current contents of the file at the given path. Returns "" (empty string view) if file does not yet
```

Comment:
Good question. The `require` statement question isn't something that is handled by Sorbet; you hand Sorbet a set of files in the workspace and it builds up the global namespace.

What this method deals with is file URIs received from the client in `textDocument/didChange` events. By the standard, a `didChange` event should always be preceded by a `didOpen` event that contains the full text of the file. The reason being is that `didChange` supports diffs from the last-known state of the file; we can't know the last known state of the file if we didn't at some point receive the whole file. And we can't go to the disk for the whole file contents, as it might be a new file that hasn't been saved to disk.

So what you're seeing here is the following: The code right now treats a `didChange` to a file that wasn't previously opened as if the file contains nothing. This is the same behavior that we had prior to this change.

I think the _correct_ thing to do would be to send back an error somehow.